tensorboard/data/server/run.rs

```diff
@@ -0,0 +1,363 @@
+        *start_time = Some(wall_time);
+    }
+    match e.what {
+        Some(pb::event::What::GraphDef(_)) => {
+            // TODO(@wchargin): Handle run graphs.
+            eprintln!("graph_def events not yet handled");
+        }
+        Some(pb::event::What::Summary(sum)) => {
+            for mut summary_value in sum.value {
+                let value = match summary_value.value {
+                    None => continue,
+                    Some(v) => v,
+                };
+
+                use std::collections::hash_map::Entry;
+                let ts = match time_series.entry(Tag(summary_value.tag)) {
+                    Entry::Occupied(o) => o.into_mut(),
+                    Entry::Vacant(v) => {
+                        let metadata = crate::data_compat::initial_metadata(
+                            summary_value.metadata.take(),
+                            &value,
+                        );
+                        v.insert(TimeSeries::new(metadata))
+                    }
+                };
+                let sv = StageValue {
+                    wall_time,
+                    payload: StagePayload::Summary(Box::new(value)),
+                };
+                ts.rsv.offer(step, sv);
+            }
+        }
+        _ => {}
+    }
+}
+
+impl Default for RunLoader {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::fs::File;
+    use std::io::{BufWriter, Write};
+
+    /// Writes an event to the given writer, in TFRecord form.
+    fn write_event<W: Write>(writer: W, event: &pb::Event) -> std::io::Result<()> {
+        use prost::Message;
+        let mut data = Vec::new();
+        event.encode(&mut data)?;
+        crate::tf_record::TfRecord::from_data(data).write(writer)?;
+        Ok(())
+    }
+
+    /// Writes a TF 1.x scalar event (`simple_value`) to the given writer.
+    fn write_scalar<W: Write>(
+        writer: W,
+        tag: &Tag,
+        step: Step,
+        wt: WallTime,
+        value: f32,
+    ) -> std::io::Result<()> {
+        let event = pb::Event {
+            step: step.0,
+            wall_time: wt.into(),
+            what: Some(pb::event::What::Summary(pb::Summary {
+                value: vec![pb::summary::Value {
+                    tag: tag.0.clone(),
+                    value: Some(pb::summary::value::Value::SimpleValue(value)),
+                    ..Default::default()
+                }],
+                ..Default::default()
+            })),
+            ..Default::default()
+        };
+        write_event(writer, &event)
+    }
+
+    #[test]
+    fn test() -> Result<(), Box<dyn std::error::Error>> {
+        let logdir = tempfile::tempdir()?;
+        let f1_name = logdir.path().join("tfevents.123");
+        let f2_name = logdir.path().join("tfevents.456");
+        let mut f1 = BufWriter::new(File::create(&f1_name)?);
+        let mut f2 = BufWriter::new(File::create(&f2_name)?);
```

Comment:
No action required.

Interesting; I thought the module would consider an inevitable abstraction of the file system and operate on testable EventFile or something akin to that. 

I assume this is actually writing these events out to the temp directory (at `sync_all` call in L315)?

Reply:
That was certainly my plan, to parameterize the reader:

```rust
pub struct RunLoader<R> {
    files: HashMap<PathBuf, EventFile<R>>,
    // ...
}
impl<R: Read> for RunLoader { /* ... */ }
```

This works swimmingly for reading, but the problem is that `Read` is
just a stream type, and doesn’t actually give you a way to _open_ a
file. So I would have needed to define something like:

```rust
trait Filesystem {
    type File: Read + Seek;
    fn open(p: &Path) -> io::Result<Self::File>;
}

struct RealFilesystem;
impl Filesystem for RealFilesystem {
    type File = std::fs::File;
    fn open(p: &Path) -> io::Result<Self::File> { std::fs::File::open(p) }
}
```

And this works perfectly fine, and we probably _will_ want it
eventually, but I didn’t really want to bother implementing my own fake
in-memory file system when the real filesystem will do just fine.

> I assume this is actually writing these events out to the temp
> directory (at `sync_all` call in L315)?

Yep. The `sync_all` is an `fsync(2)`. Without it, the files would still
be written when they’re dropped, but here we needed to keep a reference
to them later, so this flushes the `BufWriter`’s buffer.